flair/models/biomedical_entity_linking.py

```diff
@@ -0,0 +1,988 @@
+            dimension = self.dict_dense_embeds.shape[1]
+            # to use cosine similarity, we normalize the vectors and then use inner product
+            faiss.normalize_L2(self.dict_dense_embeds)
+
+            self.dense_dictionary_index = faiss.IndexFlatIP(dimension)
+            if self.index_use_cuda:
+                self.dense_dictionary_index = faiss.index_cpu_to_gpu(
+                    faiss.StandardGpuResources(), 0, self.dense_dictionary_index
+                )
+            self.dense_dictionary_index.add(self.dict_dense_embeds)
+
+            if self.index_use_cuda:
+                # create a cpu version of the index to cache (IO does not work with gpu index)
+                cache_index_dense = faiss.index_gpu_to_cpu(self.dense_dictionary_index)
+                # cache dictionary
+                self.cache_dictionary(
+                    cache_index_dense, cache_folder, cached_dictionary_path
+                )
+            else:
+                self.cache_dictionary(
+                    self.dense_dictionary_index, cache_folder, cached_dictionary_path
+                )
+
+        # apply mean centering
+        if mean_centering:
+            self.tgt_space_mean_vec = self.dict_dense_embeds.mean(0)
+            self.dict_dense_embeds -= self.tgt_space_mean_vec
+        else:
+            self.tgt_space_mean_vec = None
+
+    def load_cached_dictionary(self, cached_dictionary_path: str):
+        with open(cached_dictionary_path, "rb") as cached_file:
+            cached_dictionary = pickle.load(cached_file)
+            log.info(
+                "Loaded dictionary from cached file {}".format(cached_dictionary_path)
+            )
+
+            (self.dictionary, self.dict_sparse_embeds, self.dense_dictionary_index,) = (
+                cached_dictionary["dictionary"],
+                cached_dictionary["sparse_dictionary_embeds"],
+                cached_dictionary["dense_dictionary_index"],
+            )
+            self.dense_dictionary_index = faiss.index_cpu_to_gpu(
+                faiss.StandardGpuResources(), 0, self.dense_dictionary_index
+            )
+
+    def load_dictionary(self, dictionary_name_or_path: str):
+        # use provided dictionary
+        if dictionary_name_or_path == "ctd-disease":
+            self.dictionary = NEL_CTD_DISEASE_DICT().data
+        elif dictionary_name_or_path == "ctd-chemical":
+            self.dictionary = NEL_CTD_CHEMICAL_DICT().data
+        elif dictionary_name_or_path == "ncbi-gene":
+            self.dictionary = NEL_NCBI_GENE_DICT().data
+        elif dictionary_name_or_path == "ncbi-taxonomy":
+            self.dictionary = NEL_NCBI_TAXONOMY_DICT().data
+        # use custom dictionary file
+        else:
+            self.dictionary = DictionaryDataset(
+                dictionary_path=dictionary_name_or_path
+            ).data
+
+    def cache_dictionary(self, cache_index_dense, cache_folder, cached_dictionary_path):
+        cached_dictionary = {
```

Comment:
IMO there is no need to save "dictionary", this is just the output of reading a file w/ basically no preprocessing. For "dense_dictionary_emebds" we should use `torch.save` or whathever `flair` uses. For "sparse" there is still the question about using pickle...